File: packages/generator-core/src/css.ts
/** One CSS declaration, retaining its complete value and declaration order. */
export type StyleDeclaration = readonly [property: string, value: string];

/** Split inline CSS only at separators outside strings, comments and blocks. */
export function parseStyle(style: unknown): StyleDeclaration[] {
  if (typeof style !== 'string' || !style.trim()) return [];
  const declarations: StyleDeclaration[] = [];
  const blocks: string[] = [];
  let quote = '', escaped = false, comment = false;
  let start = 0, colon = -1;
  const finish = (end: number) => {
    if (colon >= start) {
      const property = style.slice(start, colon).replace(/\/\*[\s\S]*?\*\//g, ' ').trim();
      const value = style.slice(colon + 1, end).trim();
      if (property && value) declarations.push([property, value]);
    }
    start = end + 1;
    colon = -1;
  };
  for (let index = 0; index < style.length; index++) {
    const character = style[index]!;
    if (comment) {
      if (character === '*' && style[index + 1] === '/') { comment = false; index++; }
      continue;
    }
    if (escaped) { escaped = false; continue; }
    if (character === '\\') { escaped = true; continue; }
    if (quote) { if (character === quote) quote = ''; continue; }
    if (character === '/' && style[index + 1] === '*') { comment = true; index++; continue; }
    if (character === '"' || character === "'") { quote = character; continue; }
    if (character === '(' || character === '[' || character === '{') { blocks.push(character); continue; }
    const closing = character === ')' ? '(' : character === ']' ? '[' : character === '}' ? '{' : '';
    if (closing) { if (blocks[blocks.length - 1] === closing) blocks.pop(); continue; }
    if (blocks.length) continue;
    if (character === ':' && colon === -1) colon = index;
    else if (character === ';') finish(index);
  }
  finish(style.length);
  return declarations;
}

/** Normalize declaration spacing without changing values or repeated properties. */
export function styleString(style: unknown): string | undefined {
  const declarations = parseStyle(style);
  return declarations.length ? declarations.map(([property, value]) => `${property}: ${value}`).join('; ') : undefined;
}

File: packages/generator-rust/src/css.rs
use serde_json::{Map, Value};

pub(crate) fn declarations(source: &str) -> Vec<(String, String)> {
    if source.trim().is_empty() {
        return Vec::new();
    }
    let mut result = Vec::new();
    let mut blocks = Vec::new();
    let mut quote = None;
    let mut escaped = false;
    let mut comment = false;
    let mut skip = false;
    let mut start = 0;
    let mut colon = None;
    let comments = regex::Regex::new(r"/\*[\s\S]*?\*/").expect("CSS comment pattern");
    let finish =
        |end: usize, start: usize, colon: Option<usize>, result: &mut Vec<(String, String)>| {
            if let Some(colon) = colon {
                let property = comments
                    .replace_all(&source[start..colon], " ")
                    .trim()
                    .to_owned();
                let value = source[colon + 1..end].trim().to_owned();
                if !property.is_empty() && !value.is_empty() {
                    result.push((property, value));
                }
            }
        };
    for (index, character) in source.char_indices() {
        if skip {
            skip = false;
            continue;
        }
        if comment {
            if character == '*' && source.as_bytes().get(index + 1) == Some(&b'/') {
                comment = false;
                skip = true;
            }
            continue;
        }
        if escaped {
            escaped = false;
            continue;
        }
        if character == '\\' {
            escaped = true;
            continue;
        }
        if quote.is_some() {
            if quote == Some(character) {
                quote = None;
            }
            continue;
        }
        if character == '/' && source.as_bytes().get(index + 1) == Some(&b'*') {
            comment = true;
            skip = true;
            continue;
        }
        if character == '\'' || character == '"' {
            quote = Some(character);
            continue;
        }
        if ['(', '[', '{'].contains(&character) {
            blocks.push(character);
            continue;
        }
        let closing = match character {
            ')' => Some('('),
            ']' => Some('['),
            '}' => Some('{'),
            _ => None,
        };
        if let Some(closing) = closing {
            if blocks.last() == Some(&closing) {
                blocks.pop();
            }
            continue;
        }
        if !blocks.is_empty() {
            continue;
        }
        if character == ':' && colon.is_none() {
            colon = Some(index);
        } else if character == ';' {
            finish(index, start, colon, &mut result);
            start = index + 1;
            colon = None;
        }
    }
    finish(source.len(), start, colon, &mut result);
    result
}

pub(crate) fn style_string(source: &str) -> String {
    declarations(source)
        .into_iter()
        .map(|(property, value)| format!("{property}: {value}"))
        .collect::<Vec<_>>()
        .join("; ")
}

pub(crate) fn render_style(source: &str) -> String {
    let mut properties = Map::new();
    for (property, value) in declarations(source) {
        properties.insert(property, Value::String(value));
    }
    properties
        .iter()
        .map(|(property, value)| format!("{property}:{}", value.as_str().expect("style value")))
        .collect::<Vec<_>>()
        .join(";")
}
